### ISSUE TITLE
Secret volume plugin should be idempotent

### ISSUE THREAD

**Issue by pmorie** [GoogleCloudPlatform/kubernetes#6915]
The secret volume plugin should be idempotent and currently is not.  My suggestion for making it so is to hash the map of secret data and write the hash into the volume in something like `.secret_hash` (which will need to be a reserved value for a key in secret data).  The plugin can perform the hash, read the value from disk, and compare.  If the hashes differ, update the data on disk.

Any thoughts @thockin @vishh ?

**Comment by erictune** [comment]
When would kubelet try to mount a volume twice?

**Comment by pmorie** [comment]
@erictune Every time `syncPod` runs

**Comment by erictune** [comment]
Does this also affect any other volume types:  the proposed Config volume, Claims?  If so, is a more general mechanism needed, like a checkpoint?

**Comment by pmorie** [comment]
@erictune @thockin I believe @yifan-gu's https://github.com/GoogleCloudPlatform/kubernetes/pull/6885 PR would give us a good way to introduce a check point.

**Comment by thockin** [comment]
The simplest idempotency check is IsMountPoint(), but that doesn't verify
the data.

**Comment by thockin** [comment]
For better or worse, we assume we can do everything idempotently, and
therefore not need a checkpoint.

**Comment by thockin** [comment]
A checkpoint would trade where the complexity lives.  A nice property we
have now is that we don't have a checkpoint to lose.  If we had a
checkpoint and we intended to change the way we resync the universe, maybe
we would not need this restriction.  Yes, it affects all volumes, though
all of them are idempotent as far as I know, except Secret (which I totally
forgot to audit, mea culpa)

**Comment by thockin** [comment]
We should not adopt a checkpoint lightly - it has a lot of implications for
the kubelet codebase and performance overall.  I am not against it (in fact
I think we should) but we need to do it carefully.

**Comment by pmorie** [comment]
@thockin @erictune One thing I thought of as I went to implement this is that there are a couple approaches we could take depending on what we want.  If we want secret data to be updated on disk, we can diff the data and apply updates when there's a diff (or just write the resource version of the secret to disk and compare that), which is what I had described.  If we want the semantics to be 'one time and one time only, we will write the latest version of the secret to disk when the pod is created', we can write a marker file (`.secret_complete` or something) and check for it in `Setup`.

Taking a look at the existing plugins, the git-repo plugin writes a 'ready' file, empty dir tmpfs does a check to see whether the mountpoint already exists, etc.  I'm going to implement a 'ready' file for now.

**Comment by pmorie** [comment]
Resolved by #7166

**Comment by thockin** [comment]
Regarding updates if secrets have changed.  I am OK with that (in favor of
it, even) BUT it has to be done atomically, whcih the current
implementation can not do.  You need indirection, eg, emulate this:

```
# Nothing exists yet, create the secret mount
thockin@freakshow:/tmp/sec$ ls
thockin@freakshow:/tmp/sec$ mkdir 1
thockin@freakshow:/tmp/sec$ date > 1/date
thockin@freakshow:/tmp/sec$ whoami > 1/whoami
thockin@freakshow:/tmp/sec$ ln -s 1 current
thockin@freakshow:/tmp/sec$ ln -s current/date date
thockin@freakshow:/tmp/sec$ ln -s current/whoami whoami

# Now the secret volume is ready
thockin@freakshow:/tmp/sec$ ls -l
total 4
drwxr-x--- 2 thockin eng 4096 Apr 23 15:54 1
lrwxrwxrwx 1 thockin eng    1 Apr 23 15:54 current -> 1
lrwxrwxrwx 1 thockin eng   12 Apr 23 15:55 date -> current/date
lrwxrwxrwx 1 thockin eng   14 Apr 23 15:55 whoami -> current/whoami
thockin@freakshow:/tmp/sec$ cat date
Thu Apr 23 15:54:05 PDT 2015

# Later, the data changes
thockin@freakshow:/tmp/sec$ mkdir 2
thockin@freakshow:/tmp/sec$ date > 2/date
thockin@freakshow:/tmp/sec$ whoami > 2/whoami

# Flip to the new data atomically
thockin@freakshow:/tmp/sec$ ln -s 2 next
thockin@freakshow:/tmp/sec$ mv -T next current

# Now it is all new
thockin@freakshow:/tmp/sec$ ls -l
total 8
drwxr-x--- 2 thockin eng 4096 Apr 23 15:57 1
drwxr-x--- 2 thockin eng 4096 Apr 23 15:55 2
lrwxrwxrwx 1 thockin eng    1 Apr 23 15:56 current -> 2
lrwxrwxrwx 1 thockin eng   12 Apr 23 15:55 date -> current/date
lrwxrwxrwx 1 thockin eng   14 Apr 23 15:55 whoami -> current/whoami
thockin@freakshow:/tmp/sec$ cat date
Thu Apr 23 15:55:38 PDT 2015

# Clean up
thockin@freakshow:/tmp/sec$ rm -rf 1
```
